zebra-chain/src/sapling/tree.rs

```diff
@@ -108,71 +108,134 @@ impl From<&Root> for [u8; 32] {
+                let mut combined = merkle_crh_sapling(0, l, r);
+                self.left = Some(cm_bytes);
+                self.right = None;
+
+                for i in 0..MERKLE_DEPTH {
+                    if i < self.parents.len() {
+                        if let Some(p) = self.parents[i] {
+                            combined = merkle_crh_sapling((i + 1) as u8, p, combined);
+                            self.parents[i] = None;
+                        } else {
+                            self.parents[i] = Some(combined);
+                            break;
+                        }
+                    } else {
+                        self.parents.push(Some(combined));
+                        break;
+                    }
                 }
-                next_layer_up.push(merkle_crh_sapling(height, left, right));
             }
-
-            height += 1;
-            current_layer = next_layer_up.into();
         }
 
-        assert!(current_layer.len() == 1);
+        Ok(())
+    }
 
-        NoteCommitmentTree {
-            root: Root(current_layer.pop_front().unwrap()),
-            height,
-            count,
-        }
+    /// Returns the current root of the tree, used as an anchor in Sapling
+    /// shielded transactions.
+    pub fn root(&self) -> Root {
+        // Hash left and right together, filling in empty leaves as needed.
+        let leaf_root = merkle_crh_sapling(
+            0,
+            self.left.unwrap_or_else(|| EMPTY_ROOTS[0]),
+            self.right.unwrap_or_else(|| EMPTY_ROOTS[0]),
+        );
+
+        // Hash in parents up to the currently-filled depth, roots of the empty
+        // subtrees are used as needed.
+        let mid_root = self
+            .parents
+            .iter()
+            .enumerate()
+            .fold(leaf_root, |root, (i, p)| match p {
+                Some(node) => merkle_crh_sapling((i + 1) as u8, *node, root),
+                None => merkle_crh_sapling((i + 1) as u8, root, EMPTY_ROOTS[i + 1]),
+            });
+
+        // Hash in roots of the empty subtrees up to the final depth.
+        ((self.parents.len() + 1)..MERKLE_DEPTH)
+            .fold(mid_root, |root, d| {
+                merkle_crh_sapling(d as u8, root, EMPTY_ROOTS[d])
+            })
+            .into()
     }
-}
 
-impl NoteCommitmentTree {
     /// Get the Jubjub-based Pedersen hash of root node of this merkle tree of
-    /// commitment notes.
+    /// note commitments.
     pub fn hash(&self) -> [u8; 32] {
-        self.root.0
+        self.root().into()
+    }
+
+    /// Count of note commitments added to the tree.
+    ///
+    /// For Sapling, the tree is capped at 2^32.
+    pub fn count(&self) -> u32 {
+        self.parents.iter().enumerate().fold(
+            match (self.left, self.right) {
+                (None, None) => 0,
+                (Some(_), None) => 1,
+                (Some(_), Some(_)) => 2,
+                (None, Some(_)) => unreachable!(),
```

Comment:
Missing documentation: Can you add a panic message explaining why this case is unreachable?

```suggestion
                (None, Some(_)) => unreachable!("roots are always added left to right"),
```